src/content/1.7/development/architecture/migration-guide/validation.md

```diff
@@ -0,0 +1,18 @@
+---
+title: Validation
+weight: 70
+---
+
+# Validation
+
+The purpose of validation is to protect the application from failures, state inconsistency, security related issues etc. Prestashop has different levels of validation:
+* **Forms** (the [CRUD]({{<relref "/1.7/development/architecture/migration-guide/forms/crud-forms">}}) or [Settings]({{<relref "/1.7/development/architecture/migration-guide/forms/settings-forms">}})) - are using [Symfony validation constraints](https://symfony.com/doc/3.4/validation.html#constraints) to prevent the user from providing invalid information while also enriching the user experience. For example if a form input expects `Yes` or `No`, but user submitted value `XYZ` we want to reject the data and inform the user what was wrong. The best approach (if possible) is to prevent user providing the wrong value in first place, so in this example we could simply use a checkbox or select input with only 2 possible selections. There are some custom constraints implemented in Prestashop [here](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/src/Core/ConstraintValidator).
+* [Value objects]({{< relref "/1.7/development/architecture/domain/value_objects" >}}) - after the form is submitted and validated, controller is usually dispatching some `Command`. We must ensure that commands are valid too, because in theory, commands should not be aware whether it was dispatched from a controller or from a command line, therefore it is possible that inputs are not yet validated by the form. The `command` itself should not handle the validation (although there are some older commands that does that, but that might only be some early migration mistakes), but **use the value objects in the constructor instead**. Value objects has their own validation rules and can be reused in any other command. For example, take a look at [ProductCondition](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Core/Domain/Product/ValueObject/ProductCondition.php) - this is a value object of `Product` domain. It allows constructing specific values and throws `ProductConstraintException` if invalid one is provided. It is currently used in [UpdateProductOptionsCommand](https://github.com/PrestaShop/PrestaShop/1.7.8.x/develop/src/Core/Domain/Product/Command/UpdateProductOptionsCommand.php) and helps to maintain the command validity.
```

Comment:
```suggestion
* [Value objects]({{< relref "/1.7/development/architecture/domain/value_objects" >}}) - after the form is submitted and validated, controller is usually dispatching some `Command`. We must ensure that commands are valid too, because in theory, commands should not be aware whether they were dispatched from a controller or from a command line, therefore it is possible that inputs are not yet validated by the form. The `command` itself should not handle the validation (although there are some older commands that does that, but that might only be some early migration mistakes), but **use the value objects in the constructor instead**. Value objects have their own validation rules and can be reused in any other command. For example, take a look at [ProductCondition](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Core/Domain/Product/ValueObject/ProductCondition.php) - this is a value object of `Product` domain. It allows constructing specific values and throws `ProductConstraintException` if invalid one is provided. It is currently used in [UpdateProductOptionsCommand](https://github.com/PrestaShop/PrestaShop/1.7.8.x/develop/src/Core/Domain/Product/Command/UpdateProductOptionsCommand.php) and helps to maintain the command validity.
```